clients/PortManager/CPortManager.3
'\"
'\" Copyright (c) 2004  Michigan State University
'\" The TCL formatting macros have been shamelessly stolen
'\" from the httpd tcl man page:
'\"
'\" Copyright (c) 1995-1997 Sun Microsystems, Inc.
'\" Copyright (c) 1998-2000 by Ajuba Solutions.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
'\" The definitions below are for supplemental macros used in Tcl/Tk
'\" manual entries.
'\"
'\" .AP type name in/out ?indent?
'\"	Start paragraph describing an argument to a library procedure.
'\"	type is type of argument (int, etc.), in/out is either "in", "out",
'\"	or "in/out" to describe whether procedure reads or modifies arg,
'\"	and indent is equivalent to second arg of .IP (shouldn't ever be
'\"	needed;  use .AS below instead)
'\"
'\" .AS ?type? ?name?
'\"	Give maximum sizes of arguments for setting tab stops.  Type and
'\"	name are examples of largest possible arguments that will be passed
'\"	to .AP later.  If args are omitted, default tab stops are used.
'\"
'\" .BS
'\"	Start box enclosure.  From here until next .BE, everything will be
'\"	enclosed in one large box.
'\"
'\" .BE
'\"	End of box enclosure.
'\"
'\" .CS
'\"	Begin code excerpt.
'\"
'\" .CE
'\"	End code excerpt.
'\"
'\" .VS ?version? ?br?
'\"	Begin vertical sidebar, for use in marking newly-changed parts
'\"	of man pages.  The first argument is ignored and used for recording
'\"	the version when the .VS was added, so that the sidebars can be
'\"	found and removed when they reach a certain age.  If another argument
'\"	is present, then a line break is forced before starting the sidebar.
'\"
'\" .VE
'\"	End of vertical sidebar.
'\"
'\" .DS
'\"	Begin an indented unfilled display.
'\"
'\" .DE
'\"	End of indented unfilled display.
'\"
'\" .SO
'\"	Start of list of standard options for a Tk widget.  The
'\"	options follow on successive lines, in four columns separated
'\"	by tabs.
'\"
'\" .SE
'\"	End of list of standard options for a Tk widget.
'\"
'\" .OP cmdName dbName dbClass
'\"	Start of description of a specific option.  cmdName gives the
'\"	option's name as specified in the class command, dbName gives
'\"	the option's name in the option database, and dbClass gives
'\"	the option's class in the option database.
'\"
'\" .UL arg1 arg2
'\"	Print arg1 underlined, then print arg2 normally.
'\"
'\" RCS: @(#) $Id$
'\"
'\"	# Set up traps and other miscellaneous stuff for Tcl/Tk man pages.
.if t .wh -1.3i ^B
.nr ^l \n(.l
.ad b
'\"	# Start an argument description
.de AP
.ie !"\\$4"" .TP \\$4
.el \{\
.   ie !"\\$2"" .TP \\n()Cu
.   el          .TP 15
.\}
.ta \\n()Au \\n()Bu
.ie !"\\$3"" \{\
\&\\$1	\\fI\\$2\\fP	(\\$3)
.\".b
.\}
.el \{\
.br
.ie !"\\$2"" \{\
\&\\$1	\\fI\\$2\\fP
.\}
.el \{\
\&\\fI\\$1\\fP
.\}
.\}
..
'\"	# define tabbing values for .AP
.de AS
.nr )A 10n
.if !"\\$1"" .nr )A \\w'\\$1'u+3n
.nr )B \\n()Au+15n
.\"
.if !"\\$2"" .nr )B \\w'\\$2'u+\\n()Au+3n
.nr )C \\n()Bu+\\w'(in/out)'u+2n
..
.AS Tcl_Interp Tcl_CreateInterp in/out
'\"	# BS - start boxed text
'\"	# ^y = starting y location
'\"	# ^b = 1
.de BS
.br
.mk ^y
.nr ^b 1u
.if n .nf
.if n .ti 0
.if n \l'\\n(.lu\(ul'
.if n .fi
..
'\"	# BE - end boxed text (draw box now)
.de BE
.nf
.ti 0
.mk ^t
.ie n \l'\\n(^lu\(ul'
.el \{\
.\"	Draw four-sided box normally, but don't draw top of
.\"	box if the box started on an earlier page.
.ie !\\n(^b-1 \{\
\h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.el \}\
\h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.\}
.fi
.br
.nr ^b 0
..
'\"	# VS - start vertical sidebar
'\"	# ^Y = starting y location
'\"	# ^v = 1 (for troff;  for nroff this doesn't matter)
.de VS
.if !"\\$2"" .br
.mk ^Y
.ie n 'mc \s12\(br\s0
.el .nr ^v 1u
..
'\"	# VE - end of vertical sidebar
.de VE
.ie n 'mc
.el \{\
.ev 2
.nf
.ti 0
.mk ^t
\h'|\\n(^lu+3n'\L'|\\n(^Yu-1v\(bv'\v'\\n(^tu+1v-\\n(^Yu'\h'-|\\n(^lu+3n'
.sp -1
.fi
.ev
.\}
.nr ^v 0
..
'\"	# Special macro to handle page bottom:  finish off current
'\"	# box/sidebar if in box/sidebar mode, then invoked standard
'\"	# page bottom macro.
.de ^B
.ev 2
'ti 0
'nf
.mk ^t
.if \\n(^b \{\
.\"	Draw three-sided box if this is the box's first page,
.\"	draw two sides but no top otherwise.
.ie !\\n(^b-1 \h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.el \h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.\}
.if \\n(^v \{\
.nr ^x \\n(^tu+1v-\\n(^Yu
\kx\h'-\\nxu'\h'|\\n(^lu+3n'\ky\L'-\\n(^xu'\v'\\n(^xu'\h'|0u'\c
.\}
.bp
'fi
.ev
.if \\n(^b \{\
.mk ^y
.nr ^b 2
.\}
.if \\n(^v \{\
.mk ^Y
.\}
..
'\"	# DS - begin display
.de DS
.RS
.nf
.sp
..
'\"	# DE - end display
.de DE
.fi
.RE
.sp
..
'\"	# SO - start of list of standard options
.de SO
.SH "STANDARD OPTIONS"
.LP
.nf
.ta 5.5c 11c
.ft B
..
'\"	# SE - end of list of standard options
.de SE
.fi
.ft R
.LP
See the \\fBoptions\\fR manual entry for details on the standard options.
..
'\"	# OP - start of full description for a single option
.de OP
.LP
.nf
.ta 4c
Command-Line Name:	\\fB\\$1\\fR
Database Name:	\\fB\\$2\\fR
Database Class:	\\fB\\$3\\fR
.fi
.IP
..
'\"	# CS - begin code excerpt
.de CS
.RS
.nf
.ta .25i .5i .75i 1i
..
'\"	# CE - end code excerpt
.de CE
.fi
.RE
..
.de UL
\\$1\l'|0\(ul'\\$2
..
'\"  End of stolen macros --------------------------------
.TH "CPortManager class" 3 0.1 NSCLDaq "NSCL DAQ C++ packages"
.BS
.SH NAME
CPortManager \- C++ class interface to the Port Manager server.
.SH SYNOPSIS
.DS
#include <config.h>
#include <CPortManager.h>
#include <CPortManagerException.h>

class CPortManager {
public:
	//! portInfo lists the information about a single port.
	typedef struct _portInfo {
		int         s_Port;
		string      s_Application;
		string      s_User;
	} portInfo;
public:   
    CPortManager(string host = string("localhost"));
    CPortManager(string host, int Port);
    ~CPortManager();

    int operator==(const CPortManager& rhs) const;
    int operator!=(const CPortManager& rhs) const;

    int allocatePort(string application);
    vector<portInfo> getPortUsage();
};

.DE

add
.DS
-I/usr/opt/daq/current/include 
.DE
to your compilation lines.

add 
.DS
-L/usr/opt/daq/current/lib -lPortManager
.DE
to your link line.
.SH DESCRIPTION
.PP
\fBCPortManager\fR provides a C++ application programming interface into the
NSCL port manager server.  This class serves as a proxy for communication with
port management servers.  Once you have created a \fBCPortManager\fR object,
you can use it to allocate ports (if the server is local), and to list port
usage of any server on the network.
.PP
Note that members of this class can throw exceptions of type
\fBCPortManagerException\fR.  See the man page on that class for more information.
.SH "PUBLIC INTERFACE"
.TP
\fBCPortManager::CPortManager\fR(string \fIhost\fR = string("localhost"));
.TP
\fBCPortManager::CPortManager\fR(string \fIhost\fR, int \fIport\fR)
 Constructs a port manager object on the specified host and port.  In the first
form of the call, the class first attempts to read the port number from
/var/tmp/daqportmanager/listen.port before falling back to the default port
number of 30000.  The \fIhost\fR parameter is the name of the host to which the
object will connect when required to do so.
.TP
int \fBCPortManager::allocatePort\fR(string \fIapplication\fR);
Contacts the server and attempts to allocate a port.  If successful, the port
number is returned.  On failure a \fBCPortManagerException\fR is thrown
describing why the call failed. 
.sp
The \fIapplication\fR parameter is the name of the application for which the
port is requested.   This should be a string that clearly identifies the
function of the application.  
.sp
If a port was successfully allocated, the object will maintain a connection to
the server.  When destroyed, CPortManager objects do not close this
connection. This is done to ensure the application retains ownership 
.TP
vector<CPortManager::portInfo> \fBgetPortUsage\fR();
Interacts with the server to find out how it has allocated its ports.
Returns a vector of \fBCPortManager::portInfo\fR structs.  Each element of
the vector describes a single allocated port: \fBs_Port\fR is the number of the
port. \fBs_Application\fR is a string containing the application
name. \fBs_User\fR is a string containing the user running the application.
.SH "EXAMPLES"
.PP
Allocates a port to "MyApplication"

.DS
    . . . 
CPortManager pm;
int port;
try {
   port = pm.allocatePort("MyApplication");
   ListenOnPort(port);                      // Not shown for brevity
}
catch (CPortManagerException &error) {
   cerr << "Could not allocate a port: " << error << endl;
   exit(-1);
}
   . . .
.DE
.PP
Lists to cout the port allocations on a server in ahost.nscl.msu.edu:
.DS
 . . .
CPortManager pm("ahost.nscl.msu.edu");
vector<CPortManager::portInfo> info = pm.getPortUsage();
for(int i =0; i < info.size(); i++) {
   cout << "Port " << info.s_Port 
        << " allocated to " << info.s_Application
        << " run by " << info.s_User << endl;
}
 . . .
.DE
.SH "SEE ALSO"

portAllocator(3tcl), CPortManagerException(3), DaqPortManager(8)

ChangeLog
Dec 16, 2010 
      - Fix error in CRingItem copy construction ... if there was a
	payload, the m_pCursor member was not being set to point to the
	end of that payload...leading the CRingItem target to think the
	item was empty.  This was also a problem with operator= which also
	used CRingItem::copyin where the problem lay.
      - Fix various copy in issues with CRingBuffer and subclasses.
	make all the tests pass that should pass and remove those that
	should no longer pass due to changes in the way sampling
	has been done.
Dec 20,2010
      - Fix error in URL parsing that made absolute file urls fail.
	(removed the leading /).
      - Fix error in Gui that prevented event file segments from
	showing up on the status strip.
Dec 31, 2010
      - In readout gui, make links between files symbolic so that they
	port across backup and restores in a different directory tree as
	happens during NSCL experiment closeout.
January 27, 2011
	- Fix up Experiment.tcl so things work on etch systems (for now).
	- Fix errors in copy construction in daq/format.
February 17, 2011	
	- Fix error in event file overwite handling in readoutgui:
	  the softlink must be deleted before being made.
	
March 10, 2011 10.0-004
	- Make ring master wait until the port manager is actually alive
	rather than dying if it isn't.
	- Install VERSION file in make install target.
	- Misc. changes to Makefile.am's and configure.ac so that
	autoreconf isn't so bitchy.
April 8 2011 10.0-005
	- Stop elapsed run timer in readout gui when the run ends.
April 13, 2011 10.0-006
	- Really get the elapsed run timer logic right ...with
	interpolation from the scaler buffers.
	
June 10, 2011 10.0-007
	- Add -textheight switch Readout shell to support running
	  in on displays with limited heights.
	
June 22, 2011 10.0-008
	- Add --deleteonexit switch to stdintoring
	- CRemoteAccess - modify to start stdintoring with --deleteonexit
	  so that the proxy ring is deleted if the remote part of the
	  hoisting pipeline dies.
July 8, 2011 10.0-009
	- Extend length of getpwbyuid buffer in CReadoutMain.cpp it was
	only marginally good for some people with small gecos fields.
August 5, 2011 10.0-010
	- Remove extraneous -lpub from daqstart's Makefile.am.
October 12, 2011 10.0-011
	Fixes for defects in readout framework found when getting CAESAR ready:
	- SBS Makefile did not honor USERCXXFLAGS, USERCCFLAGS nor
 	  USERLDFLAGS support was added in a way that ensures that
	  USERCCFLAGS and USERCXXFLAGS may differ but nonetheless
	  factoring out the common flagsets.
	- Add -D's  to the Makefiel includes so that users accessing VME
	  classes get the fact that we have a mapped interface.
	- Add missing #define __CREADOUTMAIN_H to CReadoutMain.h
	- Fix error in CAENV262Busy - defined GoReady rather than GoClear
	  so it became an abstract class.  Fixed while retaining the
	  GoReady function because we know the workaround kludge/fix won't
	  disappear.
	- Call clear on main scaler bank in CExperiment after the readout
	  so scaler don't continuously increment
	* Work around issues with the documentation getting too big for 
	  docbook2dvi (it segfaults)...by generating man pages and 
	  htmldocs only.
        - Turn off driver tracing in CVMEInterface::Read. oops
	- Lock/unlock VME Interface in trigger main loop for some dwell
	  time.
	- Fix broken -p switch in tcl server.
	- Clear scalers and readout hardware much later in the
 	  initialization game.
	- Fix time handling for scaler reads.
October 17, 2011 10.0-012
	- Move begin run emit to just prior to starting the 
	  atual run itself...so we're not timing the hardware
	  initialization.

October 18, 2011 10.0-013	
	- Stop using run state time for anything but keep computing each
	from 'now' and start stamp.
October 18, 2011 10.0-014
	- Use ns resolutino realtime clock for all CExperiment timing
	to try to get better synch with the scaler trigger clock.
	- Fix error in CRingSelectionPredicate...would not allow the last
	ring item to be fetched until more were in the ring.
October 18, 2011 10.0-015
	- Pull in up to date scaler display program from 8.2..which
	diverged after this forked off.
October 21, 2011 10.0-016
	- Fix Packet/Vpacket macros so that they generate compatibility
	  (word size-field) packets not checking for jumbo-bufferness.
October 24, 2011 10.0-017
	- Add memory barriers on shm operations that write to ensure
	everthing is flushed to shm. - gcc specific code.
	- Add a bit of a safety zone between get/put pointers.
	- Go to a two phase client allocation: first set pid - 0
	  which makes it ineligible for size computation, and claims it
	  for use.  Then set the get pointer and finally set the PID to
	  the actual value making it elligible for size computation.
Nobvember 4, 2011 10.0-018
	- Deal a bit better with pointer skew that can happen as new
	  consumers enter the fray.
	- Play with the tkdumper help file...fix missing link.
	- Provide migration tutorial.
	
November 4, 2011 10.0-019
	- Make stdintoring understand that data are structured not a
	stream.
November 7, 2011 10.0-020
	- If a proxy ring has no producer then restart the pipeline if a
	client attaches.  This provides a way to recover from remote
	system reboot that's much saner.
        - If a ring is deleted then kill its clients.
	- If a get/put/peek operations discoveres the client's PID has
	been co-opted by some other process or released, throw a state
	exception and refuse to perform the operation.  This deals with an
	edge case that should not happen...however better safe than sorry
	especially since the cost is negligible.
	- Require stdintoring have at least a full header in putData
	else the remainder is a residual.
	- Use same run time offset algorithm as scalers for variable dump
	bufs.
November 9-16, 2011 10.0-021
	- Add --exitonend option to ringselector (WO W00957).
	- Add compatibilitybuffer - ring to daq8.2 buffer utility (W00957)

clients/PortManager/CPortManager.cpp
//  This file contains the implementation of the CPortManager class
//  This class interacts with TCP port allocation servers in the NSCL
//  DAQ system.  For more information about this class and its interfaces
//  see CPortManager.h
//
// Author:
//   Ron Fox
//   NSCL
//   Michigan State University
//   East Lansing, MI 48824-1321
//
// (c) 2005 Copyright Board of Trustees of Michigan State University All rights Reserved
// This software is licensed under the Gnu Public License  For license
// terms see: http://www.gnu.org/licenses/gpl.txt
//

#include <config.h>
#include "CPortManager.h"
#include "CPortManagerException.h"


// headers needed for the network.

#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet/ip.h>

#include <assert.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <pwd.h>
#include <regex.h>
#include <stdlib.h>
#include <publib.h>
#include <ctype.h>

#include <Fstream.h>
#include <Sstream.h>

#ifdef HAVE_STD_NAMESPACE
using namespace std;
#endif

//  consts:

static const int DefaultPort(30000);      //!< If there is no listen.port file.

static const string LocalServerPortFile("/var/tmp/daqportmanager/listen.port");

// ---------------------------------------------------------------------------

/*!
 * Construct a CPortManager.  The port manager will not connect on constrution.
 * Connections will be on an as needed basis until the first successful port
 * allocation at which time the object will hold open a connection to the server
 * as required by the protocol.  
 * 
 * \param host (string [in] = "localhost")
 *     The host this port manager will interact with. 
 * \param port (int [in]):
 *     The port to connect with.  This defaults as follows, in the first
 *     form of the constructor:
 *     - If possible, the file /var/tmp/daqportmanager/listen.port is
 *       read for the port.
 *     - If /var/tmp/daqportmanager/listen.port could not be read for the port,
 *       the port defaults to DefaultPort defined above.
*/
CPortManager::CPortManager(string host) :
   m_sHost(host),
   m_nPort(DetermineDefaultPort()),
   m_fisConnected(false)
 {
 }
 CPortManager::CPortManager(string host, int port) :
 	m_sHost(host),
 	m_nPort(port),
 	m_fisConnected(false)
 {
 }
 /*!
  *  Destroy an instance of a CPortManager.  Note that contrary to what you
  * might expect, this is a no-op.  This is because there's a policy assumption
  * that the application will hold on to ports it allocated for the life of the
  * application.  In order to do this, the protocol requires that the application
  * retain an open socket to the server.  Thus, even if m_fisConnected is true,
  * we won't shutdown the socket.
  */
CPortManager::~CPortManager()
{
}

/*!
 *   Equality comparison.  Two port managers are considered equivalent if
 * they are talking (or will talk) to the same server on the same port.
 * 
 * \param rhs (const CPortManager& [in]):
 *     The port manager to compare *this to.
 * \return int
 * \retval 0  - *this is not equivalent to rhs.
 * \retval 1  - *this is equivalent to rhs.
 */
int
CPortManager::operator==(const CPortManager& rhs) const
{
	return ( (m_sHost   == rhs.m_sHost)     &&
	         (m_nPort   == rhs.m_nPort));
} 
/*!
 * Inequality comparison.  Two port managers are inequal if operator== 
 * returns false.
 */
int
CPortManager::operator!=(const CPortManager& rhs) const
{
	return !(*this == rhs);
}
// --------------------------------- class operations -----------------------

/*!
 * Allocate a port from the TCP/IP port server.
 * We connect to the server and send it a "GIMME" command.
 * Note that the user is responsible for any application name quoting
 * if the appliation name has spaces or other TCL List breaking characters.
 * 
 * We read back the server reply:
 * -  On FAIL we throw up a CPortManagerException with the reason
 *    decoded to one of NoPorts or NotLocal depending on the actual error
 *    message gotten from the server.
 * - On OK we decode the port and pass that back to the caller.
 * 
 * \param application (string [in]):
 *    The name of an application to identify what the port will be used for.
 *    This must be a 'valid' TCL list element.  e.g.:
 *    - SpecTcl
 *    - {This is a test}
 * \return int
 * \retval  The port that was allocated by the server.
 */
int
CPortManager::allocatePort(string application)
{
	Connect();
	
	// Build up the command and send it.
	
	string command("GIMME ");
	command    += application;
	command    += " ";
	command    += GetUsername();
	command    += '\n';
	
	if(send(m_nSocket, command.c_str(), command.size(), 0) != command.size()) {
		int status = errno;
		
		m_fisConnected = false;        // We've probably lost our connection.
		Disconnect();
		
		string doing(" Sending: ");
		doing +=  command;
		doing += " to the port manager server : ";
		doing += strerror(status);
		throw CPortManagerException(m_sHost,
									CPortManagerException::ConnectionFailed,
									doing);
		
	}
	
	// Get the reply: 
	
	string reply = GetLine();
	
	// Simplest decode is with an istringstream:
	
	istringstream s(reply);
	string status;
	
	s >> status;
	if(status != string("OK")) {
		Disconnect();
		s >> status;
		CPortManagerException::Reason 
		         reason(CPortManagerException::ConnectionFailed); // impossible.
 		if(status.find("no free ports") != string::npos) {
			reason = CPortManagerException::NoPorts;
		}
		if(status.find("Only local applications get ports") != string::npos) {
			reason = CPortManagerException::NotLocal;
		}
		assert(reason != CPortManagerException::ConnectionFailed);
		throw CPortManagerException(m_sHost,
									reason,
									" Attempting to allocate a port");
	}
	
	 int port;
	 s >> port;
	 m_fisConnected = true;         // We'll hold the connection from now on.
	 
	 return port;
	
	
}
/*!
 *     Determine which ports have been allocated and send the user back
 *  a vector of portInfo structs.
 * 
 * \return vector<portInfo>
 *     The list of port usages.
 * \throw CPortManagerException
 *    Due to connection/communication problems.
 * */
 vector<CPortManager::portInfo>
 CPortManager::getPortUsage()
 {
 	Connect();
 	
 	string command("LIST\n");
 	
 	if(send(m_nSocket, command.c_str(), command.size(), 0) != command.size()) {
		int status = errno;
		
		m_fisConnected = false;        // We've probably lost our connection.
		Disconnect();
		
		string doing(" Sending: ");
		doing +=  command;
		doing += " to the port manager server : ";
		doing += strerror(status);
		throw CPortManagerException(m_sHost,
					    CPortManagerException::ConnectionFailed,
					    doing);
	 	}
 	
 	// Get the reply.. should be OK nlines followed by nlines of information.
 	
 	string line = GetLine();
 	istringstream s(line);
 	string status;
 	
 	s >> status;
 	if(status != string("OK")) {
 		Disconnect();
 		string doing("Getting port usage list from server : ");
 		doing += line;
 		
 		throw CPortManagerException(m_sHost,
 									CPortManagerException::ConnectionFailed,
 									doing);
 	}
 	// Rest of the line should be number of lines to read:
 	
 	int nlines;
 	s >> nlines;
 	vector<portInfo> result;
 	for(int i =0; i < nlines; i++) {
	  portInfo info;
	  GetPortInfo(info, GetLine());
	  result.push_back(info);
 	}
 	Disconnect();
 	return result;
 }
// -------------------------------- Utility operations ---------------------

/*!
  Connect to the remote system.   If m_fisConnected, this is a no-op as
  we are already connected.  Otherwise, a socket is created, the address of
  the remote system determined and we connect to the remote system.
  The connected socket will be stored in m_nSocket for use by the rest of
  the object.
  
  \throw CPortManagerException
      Will be thrown if any part of the connection process fails.
*/
void
CPortManager::Connect()
{
  if(!m_fisConnected) {
    
    // Attempt to create the socket:
    
    m_nSocket = socket(PF_INET, SOCK_STREAM, tcp());
    if(m_nSocket < 0) {
      int stat = errno;              // String creation etc. can change errno
      string doing(" opening socket : ");
      doing += strerror(stat);
      
      throw CPortManagerException(m_sHost, 
				  CPortManagerException::ConnectionFailed,
				  doing);
    }
    // Translate the host and port to a sockaddr_in structure:
    
    sockaddr_in remoteAddress;
    
    try {
      GetNetworkAddress(remoteAddress);      // Throws on error.
    }
    catch (...) {
      close(m_nSocket);                       // prevent socket leaks.
      throw;
    }
    
    // connect to the server:
    
    if(connect(m_nSocket, (const sockaddr*)&remoteAddress, sizeof(remoteAddress))) {
      int    stat = errno;
      string doing("Connecting to server : ");
      doing  += strerror(stat);
      close(m_nSocket);
      throw CPortManagerException(m_sHost,
				  CPortManagerException::ConnectionFailed,
				  doing);
    }
  }
  
}
/*!
   Disconnect the socket, unless it's necessary to hold onto a persistent
   connection with the remote host.
*/
void
CPortManager::Disconnect()
{
	// m_fisConnected means holding a persistent connection.. We should only
	// call Disconnect if a socket is open.
	
	if(!m_fisConnected) {
		shutdown(m_nSocket, SHUT_RDWR);
		close(m_nSocket);                   // This may  not be needed?
	}
}
/*!
 * Determine the default port on which to connect.  This is done as follows:
 * First we try to open and read the port number from the file
 * in the const LocalServerPortFile.  If that fails, then we fall back to 
 * the server default port which is in the const DefaultPort
 * \return int
 * \retval  The port on which to connect by default.
*/
int
CPortManager::DetermineDefaultPort()
{
	ifstream portfile(LocalServerPortFile.c_str());
	
	// If we can't even open the port file fall back:
	
	if(!portfile) {
		return DefaultPort;
	}
	int port;
	
	portfile >> port;
	
	// If could not read the port, fall back
	
	if(!portfile) {
		return DefaultPort;
	}
	// Got a good port from the file return it:
	
	return port;
	
}

/*!
   Determine the network address of the peer that we will be talking to.
   this is used prior to a connection to create the sockaddr_in that
   is needed to connect(2).  The port is straightforward as it only needs
   to be converted to network byte ordering.  The host is a bit more interesting.
   - First we try to look it up via gethostbyname(3). If that's successful,
     that's what we use.
   - If the gethostbyname(3) fails, we attempt to treat the hostname as a
     dotted ip string (e.g. "11.22.33.44") and translate it using
     inet_aton(3).
   - If both of the above fail we will throw an exception to indicate the
   network address cannot be computed.  The exception thrown will be a 
   CPortManagerException with the reason ConnectionFailed.
   
   \param result (sockaddr_in& [out]):
       Buffer to hold the resulting network address.
   \throw CPortManagerException
       If the host does not translate.
       
    \note  This could be made a member function, and could cache the
      lookup after doing it once. The assumption is that there will not be
      requent interactions with the server and therefore there's no real penalty
      to looking up the hostname each time we need to connect.
    \note This implementation can only handle AF_INET .. we fail on AF_INET6.
*/
void
CPortManager::GetNetworkAddress(sockaddr_in& result) const
{
	// Do the easy stuff first.
	
	result.sin_family = AF_INET;
	result.sin_port   = htons(m_nPort);
	
	// Attempt to translate the hostname:
	
	struct hostent* pEntry = gethostbyname(m_sHost.c_str());
	if(pEntry) {
		if(pEntry->h_addrtype != AF_INET) {
			throw CPortManagerException(m_sHost,
			                            CPortManagerException::ConnectionFailed,
			                            " Host is not AF_INET ");
		}
		memcpy(&(result.sin_addr), pEntry->h_addr, pEntry->h_length);
		return;
	}
	// For whatever reason, we could not translate the host by name.
	// try by number:
	
	if(!inet_aton(m_sHost.c_str(), &(result.sin_addr))) {
		throw CPortManagerException(m_sHost,
									CPortManagerException::ConnectionFailed,
									"Host lookup failed");	
	}	
}
/*!
 * Get a line of text from the socket.  Unfortunately, the only way to do this
 * is character by character reads on the socket until the \n is seen.
 * Note that the \n is removed from the input string.
 *   Any errors in recv are treated as worth a ConnectionFailed CPortManager
 * exception
 * 
 * \return string
 * \retval  the input string gotten from the socket with the \n removed.
*/
string
CPortManager::GetLine() const
{
  string result;
  while(1) {
    char c;
    if(recv(m_nSocket, &c, 1, 0) != 1) {
      int stat = errno;
      string doing(" Reading a line of input from the server : ");
      doing += strerror(errno);
      
      throw CPortManagerException(m_sHost,
				  CPortManagerException::ConnectionFailed,
				  doing);
    }
    if(c == '\n') {
      break;
    }
    if(c != '\r') {		// Seems like tcl terminates with \r\n.
      result += c;
    }
  }
  
  return result;
}

/*!
	Break up a port description line into its components:
	- port - the port the line describes.
	- app  - The application described by the line.
	- user - The user who is running the application.
	
	\param line (string [in]):
	    The line to parse.
	\param result ( CPortManager::portInfo [out]):
	   The elements of the portInfo struct are filled in with the parsed
	        results.

	\note For now, if any of the items required brace quoting the braces will
	be returned with the strings... that is the line is just broken up into
	three chunks:  The first chunk is a leading number terminated by the
	first instance of whitespace.  The third chunk is a trailing set of 
	characters that is terminated by the last whitespace character.
	The third chunk is the stuff in the middle with leading and trailing 
	whitespace stripped off.
*/
void
CPortManager::GetPortInfo(CPortManager::portInfo& result, string l)
{
  string line(l);
  
  // Regular expressions we will use:
  
  string leadingnumber("^[0-9]+");
  string trailingusername("([a-z])([a-z,0-9]*)$");
  
  regex_t lnumber;
  regex_t tusername;
  
  assert(regcomp(&lnumber, leadingnumber.c_str(), REG_EXTENDED) == 0);
  assert(regcomp(&tusername, trailingusername.c_str(), REG_EXTENDED) == 0);
  
  // Get the port number extent and convert the substring to the port.
  
  regmatch_t portextent;
  assert(regexec(&lnumber, line.c_str(), 1, &portextent,0) == 0);
  string port;
  port.assign(line, portextent.rm_so,
	      portextent.rm_eo - portextent.rm_so);
  result.s_Port = atoi(port.c_str());
  
  // Get the username extent and convert the substring into a username.
  // we assume usernames are of the form [a-z][a-z,0-9]* 
  // and are butted right up to the end of the string.
  
  regmatch_t userextent;
  assert(regexec(&tusername, line.c_str(), 1, &userextent, 0) == 0);
  result.s_User.assign(line, userextent.rm_so,
		       userextent.rm_eo - userextent.rm_so);
  
  
  regfree(&lnumber);
  regfree(&tusername);
  // The string between the two starts is most of the application name.
  // we need to trim whitespace off the front and back of this.
  
  int start = portextent.rm_eo + 1; // string starts earliest here.
  int end   = userextent.rm_so - 1; // and ends earliest here.
  
  while(start <= end) {
    if(!isspace(line[start])) {
      break;
    }
    start++;
  }
  while(end >= start) {
    if(!isspace(line[end])) {
      break;
    }
    end--;
  }
  result.s_Application.assign(line, start, end-start+1);
  
}
/*!  Get and return the name of the user that corresponds to the current
  effective userid.
  
  \return string
  \retval the name of the user running.
*/
string
CPortManager::GetUsername()
{
  uid_t uid = geteuid();
  passwd* p = getpwuid(uid);
  
  if(!p) {
    return string("*unknown*");
  }
  return string(p->pw_name);
}
/*!
  Return the Tcp protocol number.
  If this cannot be found, the function will assert fail.
*/
int
CPortManager::tcp()
{
  struct protoent* pProto = getprotobyname("tcp");
  assert(pProto);
  return pProto->p_proto;
}
